### ISSUE TITLE
Components 'loaded' Promise & loopback <--> components contract 

### ISSUE THREAD

**Issue by anodynos** [strongloop/loopback#3248]
New Feature: 

It would be great if we could add support for a Promise to be returned for each `loopback-component-XXX` loaded status + more. 

This will allow us to decide whether to perform a task or not, or even disable parts of the API's functionality, wait before running tests etc. 

See for example this ~~workaround~~ beginning of a nice and simple loopback <--> components contract for auto-migration https://github.com/snowyu/loopback-component-auto-migrate.js/pull/4

In summary, we need a (lightweight) contract so components can be looked up as `app.components['superPlugin']`, and provide some rudimentary lifecycle (`then`) hooks (eg `loaded`, `stopped` etc) and even have a custom exported object.

**Comment by bajtos** [comment]
@raymondfeng @davidcheung could you PTAL please? How does this fit into the new loopback-boot design/architecture?

FWIW, loopback-component-explorer is "publishing" its configuration via `app.set('loopback-component-explorer')`, see https://github.com/strongloop/loopback-component-explorer/blob/188603f352e1a6a6ba7587aaa0a9e3e1a75ce387/index.js#L38
